yew-functional-macro/tests/functional_macro/bad-return-type-fail.stderr

```diff
@@ -0,0 +1,21 @@
+error[E0308]: mismatched types
+  --> $DIR/bad-return-type-fail.rs:10:26
+   |
+10 |   fn comp_1(props: &Props) {
+   |  __________________________^
+11 | |     let _x = html! {
+12 | |         <p>
+13 | |             { props.a }
+14 | |         </p>
+15 | |     };
+16 | | }
+   | |_^ expected enum `yew::virtual_dom::vnode::VNode`, found `()`
+
+error[E0308]: mismatched types
+  --> $DIR/bad-return-type-fail.rs:25:5
+   |
+18 | #[functional_component(Comp)]
+   | ----------------------------- expected `yew::virtual_dom::vnode::VNode` because of return type
+...
+25 |     1
+   |     ^ expected enum `yew::virtual_dom::vnode::VNode`, found integer
```

Comment:
Not perfectly happy with this error message. Should we do something about it?

Reply:
What can be done about it, considering that checking return type is impossible?

Reply:
Well the issue here is that the span points at the macro invocation (i.e. call site). If we give the return type of the trait method the same span as the input function return type it will point there instead.

Reply:
How would that be done?

Reply:
First, extract the return type from the signature. As an additional benefit we can now also provide a custom error message for when the input function doesn't have a return type at all.
```rust
let return_ty = match sig.output {
    ReturnType::Default => {
        return Err(syn::Error::new_spanned(sig, "function must return `Html`"));
    }
    ReturnType::Type(_, ty) => ty,
};
```

And in the output we apply the span from the input return type to our `Html`:
```rust
let return_ty = quote_spanned!(return_ty.span()=> ::yew::html::Html);

let quoted = quote! {
    #vis struct #function_name;

    impl ::yew_functional::FunctionProvider for #function_name {
        type TProps = #props_type;

        fn run(#arg) -> #return_ty {
            #body
        }
    }

    #vis type #component_name = ::yew_functional::FunctionComponent<#function_name>;
};
```

The new error will look like this:
```rust
error[E0308]: mismatched types
  --> $DIR/bad-return-type-fail.rs:25:5
   |
19 | fn comp(props: &Props) -> u32 {
   |                           --- expected `yew::virtual_dom::vnode::VNode` because of return type
...
25 |     1
   |     ^ expected enum `yew::virtual_dom::vnode::VNode`, found integer
```
That's pretty nice, isn't it?

Reply:
That is indeed nice. Fixed that.

Reply:
You ignored the first part of my suggestion. The error spans are significantly worse because of that.

Reply:
You missed another thing for the `ReturnType::Default` case. You're using the span from `sig.output` but that doesn't have a span because there is no output type, so it falls back to call site. If you look at my suggestion again, you'll see that I'm using `sig` for the span. That way the error points at the `fn comp()` part which is probably the best we can do.

Reply:
Oops, proc macros really are fragile.